src/clj/fluree/db/reasoner.cljc

```diff
@@ -0,0 +1,276 @@
+
+(defn schedule
+  "Returns list of rule @id values in the order they should be run.
+
+  If optional result-summary is provided, rules that don't need to be
+  re-run will be filtered out.
+
+  A result summary is list/set of the rule dependency patterns which
+  match newly created Flakes from the last run. When the result-summary
+  is empty, no rules will be left to run, but based on the dependencies
+  it is possible no rules will be left to run even if the result-summary
+  is non-empty"
+  ([rules]
+   (task-queue rules))
+  ([rules result-summary]
+   (task-queue rules result-summary)))
+
+(defn reasoner-insert
+  "When triples from rules require explicit inserts, returns flakes."
+  [db fuel-tracker txn-id author-did rule-id insert-smt]
+  (go-try
+    (let [tx-state      (-> (transact/->tx-state db txn-id author-did rule-id)
+                            (assoc :stage-update? true))
+          flakes-ch     (transact/generate-flakes db fuel-tracker insert-smt tx-state)
+          fuel-error-ch (:error-ch fuel-tracker)
+          chans         (remove nil? [fuel-error-ch flakes-ch])
+          [flakes] (alts! chans :priority true)]
+      (if (util/exception? flakes)
+        (throw flakes)
+        (let [[db* new-flakes] flakes
+              tx-state* (assoc tx-state :stage-update? true)
+              {:keys [db-after add]} (transact/final-db db* new-flakes tx-state*)]
+          [db-after add])))))
+
+(defn reasoner-stage
+  [db fuel-tracker txn-id author-did rule-id full-rule]
+  (go-try
+    (let [tx-state   (transact/->tx-state db txn-id author-did rule-id)
+          parsed-txn (:rule-parsed full-rule)
+          _          (when-not (:where parsed-txn)
+                       (throw (ex-info (str "Unable to execute reasoner rule transaction due to format error: " (:rule full-rule))
+                                       {:status 400 :error :db/invalid-transaction})))
+          flakes     (async/<! (transact/generate-flakes db fuel-tracker parsed-txn tx-state))]
+      (when (util/exception? flakes)
+        (throw flakes))
+      flakes)))
+
+;; TODO - can remove when filter function works for use case
+(defn filter-same-as-trans
+  "Note - this remove 'self' from sameAs transitive
+  rule. This would not be necessary if the filter function
+  allowed you to filter 'o' values that are equal to 's' values
+  but until that works. this addresses the issue."
+  [rule-id new-flakes]
+  (if (str/starts-with? rule-id (str const/iri-owl:sameAs "(trans)"))
+    (reduce
+      (fn [acc new-flake]
+        (if (= (flake/s new-flake) (flake/o new-flake))
+          acc
+          (conj acc new-flake)))
+      (empty new-flakes)
+      new-flakes)
+    new-flakes))
+
+(defn execute-reasoner-rule
+  [db rule-id reasoning-rules fuel-tracker {:keys [txn-id author-did] :as tx-state}]
+  (go-try
+    (let [[db reasoner-flakes] (<? (reasoner-stage db fuel-tracker txn-id author-did rule-id (get reasoning-rules rule-id)))
```

Comment:
There are some unclear dependencies in tx-state so did this for now to stay safe. I think a review of tx-state as a whole is warranted, and perhaps a better mechanism (or better docs on what is used where) would be beneficial.

Reply:
That makes sense. Agreed. I think [Dan's shacl pr](https://github.com/fluree/db/pull/755) makes more changes to tx-state by getting rid of the subj-mods atom. I had also been planning on trying to identify places we could speed up transactions as part of the large data load tests that Andrew and I have been working on for a while, and tx-state could limit the performance gains we could get by increasing parallelization by increasing buffer sizes in async channels (although it's debatable how much more performance we could get with that strategy because a lot of this is cpu bound).